### PR TITLE
chore(gatsby): Migrate utils/get-public-path.js to ts

### PR DESCRIPTION
**Pull request by cola119**
## Description
Convert `utils/get-public-path` to typescript

## Related Issues
Related to #21995

### CODE REVIEW

File: packages/gatsby/src/utils/get-public-path.ts [hunk 1]
```diff
@@ -1,16 +1,20 @@
...
-}) {
+}: {
+  assetPrefix?: string
+  pathPrefix?: string
+  prefixPaths: boolean
+}): string => {
   if (prefixPaths && (assetPrefix || pathPrefix)) {
     const normalized = [assetPrefix, pathPrefix]
-      .filter(part => part && part.length > 0)
+      .filter((part): part is string => (part ? part.length > 0 : false))
```
Comment (blainekasten) [inline packages/gatsby/src/utils/get-public-path.ts]: `part is string` is this valid syntax? Why not just `part: string` I think this line is wrong. I think you want `(part: string): boolean => ..`

**Comment by cola119** [comment]
> since we are changing the export signature there are a handful of files that will need to have their require statement updated

Sorry I missed it. I will fix that right away.

File: packages/gatsby/src/utils/get-public-path.ts (hunk 1 above)
Comment (cola119) [reply packages/gatsby/src/utils/get-public-path.ts]: `part is string` is valid syntax called [type predicate](https://www.typescriptlang.org/docs/handbook/advanced-types.html#using-type-predicates) officially. As far as I know, TS cannot infer types exactly inside `Array.prototype.reduce()` so if it is `(part: string): boolean => ..` the compiler infers `part` in the next `map` as `string | undefined`
[example](http://www.typescriptlang.org/play/#code/MYewdgzgLgBFBOBLAtgZQDYEMIAsCmEMAvDABQAOm8UAXDNEmAOYCUdDizxAfDJdQDp4eclmB5SAelIA9ADqSWAH1IKAJC0lMANDAAGelgCgjoSLEQQAqgCUAMsTLkQECIgBG6PFfjp2CTlY6dxAQL0wwHiMYGABtPRwoKHIaSUk9XQSk8ghU9My0vQBdAQgQZAk8AA9yeB4+FzdPb19SqCooCAB1RCgcUmralmMjQZBqGDNoGCY8KAAFAFdPRGB5zD7HUgBvaJhsCDn54QAzRCrtPco+47wzi6vT8-W+iEuAXzpdmIOjp6qAPz+RhMK4bHC3e5A+gBZiPO7PcG5GAhMJ4CJGd5sGEg+rfGCIE5Of4vHCEABk5LIvwW-xgSiUfHBkPOwxg+JiU1gYHGyEw6EQAC88AATRyxGksi5Mm7-Ip7GIxARndBQPDwUgUDrY1HhSJEXhaiYApmCLzMTa8AAMMDoJ35h2GCsVAj55CNsANcCQaCwuAIHrZaRgAB4ALQR020HGBekwRZgEUIsCi51KgBWIE4pD06RGipgwigi3gkUstjspB58D5AuFIpYMBN1drQtFtv0kjU2xb-LbIveej27xMMSLJciBkxQA)

Comment (blainekasten) [reply packages/gatsby/src/utils/get-public-path.ts]: TIL! V cool, thanks!

**Pull request merged**